File: java-example/BDD.java
package com.example.username;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertFalse;

public class BDDTest {
    
    private UsernameValidator validator;
    
    @BeforeEach
    void setUp() {
        validator = new UsernameValidator();
    }
    
    @Test
    void shouldAcceptValidUsernames() {
        // Examples of valid usernames
        assertTrue(validator.isValid("abc"));
        assertTrue(validator.isValid("user123"));
        assertTrue(validator.isValid("valid_username"));
    }

    @Test
    void shouldRejectTooShortUsernames() {
        // Examples of too short usernames
        assertFalse(validator.isValid(""));
        assertFalse(validator.isValid("ab"));
    }

    @Test
    void shouldRejectTooLongUsernames() {
        // Examples of too long usernames
        assertFalse(validator.isValid("abcdefghijklmnopqrstuvwxyz"));
    }

    @Test
    void shouldRejectUsernamesWithIllegalChars() {
        // Examples of usernames with illegal chars
        assertFalse(validator.isValid("user@name"));
        assertFalse(validator.isValid("special$chars"));
    }
}

File: javascript-example/BDD.js
const assert = require("assert");
const UsernameValidator = require("./UsernameValidator");

describe("Username Validator BDD Style", () => {
  let validator;

  beforeEach(() => {
    validator = new UsernameValidator();
  });

  it("should accept valid usernames", () => {
    // Examples of valid usernames
    assert.strictEqual(validator.isValid("abc"), true);
    assert.strictEqual(validator.isValid("user123"), true);
    assert.strictEqual(validator.isValid("valid_username"), true);
  });

  it("should reject too short usernames", () => {
    // Examples of too short usernames
    assert.strictEqual(validator.isValid(""), false);
    assert.strictEqual(validator.isValid("ab"), false);
  });

  it("should reject too long usernames", () => {
    // Examples of too long usernames
    assert.strictEqual(validator.isValid("abcdefghijklmnopqrstuvwxyz"), false);
  });

  it("should reject usernames with illegal chars", () => {
    // Examples of usernames with illegal chars
    assert.strictEqual(validator.isValid("user@name"), false);
    assert.strictEqual(validator.isValid("special$chars"), false);
  });
});
